events.cpp

```diff
@@ -0,0 +1,125 @@
+// START A3HEADER
+//
+// This source file is part of the Atlantis PBM game program.
+// Copyright (C) 2020 Valdi Zobēla
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program, in the file license.txt. If not, write
+// to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+// Boston, MA 02111-1307, USA.
+//
+// See the Atlantis Project web page for details:
+// http://www.prankster.com/project
+//
+// END A3HEADER
+
+#include "gameio.h"
+#include "events.h"
+
+BattleFact::~BattleFact() {
```

Comment:
Do you want to have multiple different structures for every type of event?

Reply:
This is a fact, not yet an event. Its task is to collect read-only facts about something that happened during the game. After that facts will be transformed into 0..n events. The event data structure is fixed: type, score, and event text. The score will be used to rank events among other events from the same category and the engine will take N top events from each category and put them into Times.

Reply:
Therefore each fact will have only the information it needs. This allows fast and easy extensions of this system with new facts and events later.